aten/src/ATen/native/xnnpack/MaxPooling.cpp

```diff
@@ -0,0 +1,212 @@
+#ifdef USE_XNNPACK
+
+#include <ATen/native/Pool.h>
+#include <ATen/native/xnnpack/Common.h>
+#include <ATen/native/xnnpack/Factory.h>
+#include <ATen/native/xnnpack/Pooling.h>
+
+namespace at {
+namespace native {
+namespace xnnpack {
+
+// Supports NHWC and NCHW FP32 max pooling with any
+//  - kernel size
+//  - padding
+//  - stride
+//  - dilation
+
+bool use_max_pool2d(
+    const Tensor& input,
+    const IntArrayRef kernel_,
+    const IntArrayRef padding_,
+    IntArrayRef stride_,
+    const IntArrayRef dilation_,
+    const bool ceil_mode,
+    const float output_min,
+    const float output_max) {
+  using namespace internal;
+
+  // Make sure we are not dealing with an unorthodox configuration.
+  if (kernel_.empty() || padding_.empty() || dilation_.empty()) {
+    return false;
+  }
+
+  // Stride can be legitimately empty, in which case it is to be defaulted to kernel size.
+  if (stride_.empty()) {
+    stride_ = kernel_;
+  }
+
+  // Normalize the parameters.
+  const internal::pooling::Parameters parameters{
+    kernel_,
+    padding_,
+    stride_,
+    dilation_,
+  };
+
+  // Here are the list of conditions required for this code path to be taken:
+  // * Input must be 4D CPU float tensor with no gradients.
+  // * Kernel must be a 2D IntArrayRef containing two positive numbers.
+  //   Furthermore, 1x1 kernels are neither valid nor meaningful.
+  // * Padding must be a 2D IntArrayRef containing two non-negative numbers.
+  // * Stride must be a 2D IntArrayRef containing two positive numbers.
+  // * Dilation must be a 2D IntArrayRef containing two positive numbers.
+  // * Ceil mode is not supported and must be disabled.
+  // * output_max must be greater than output_min.
+  //   Namely, setting both output_min and output_max to 0 is not valid usage.
+  // * Finally, application of this operator to the input tensor with the given
+  //   max pool 2d parameters must result in an output tensor with a valid shape.
+
+  return xnnpack::internal::available() &&
+      // Input
+      (4 == input.dim()) &&
+      (c10::DeviceType::CPU == input.device().type()) &&
+      (kFloat == input.scalar_type()) &&
+      !input.requires_grad() &&
+      // Kernel
+      (2 == parameters.kernel.size()) &&
+      (parameters.kernel[Layout::Parameter::height] > 0) &&
+      (parameters.kernel[Layout::Parameter::width] > 0) &&
+      ((parameters.kernel[Layout::Parameter::height] *
+        parameters.kernel[Layout::Parameter::width]) > 1) &&
```

Comment:
Why this condition? If kernel high and width both must be integers and > 0, then this condition is implied, no?

Reply:
1x1 max pooling is not meaningful and XNNPACK prohibits that use case.  As for the check, let's read > 0 as >= 1 to make it more intuitive.  With that in mind, for any integer M and N greater than or equal to 1, MxN, 1xN, and Mx1 are all valid sizes except for 1x1 which is what this check is guarding against.